File: 20251220 - UI_MAP_BUILDER_INTEGRATION/solution_first/pattern_complexity_modifier.py
"""
Pattern Complexity Modifier

Modifies item patterns based on difficulty level WITHOUT changing item count.
- EASY: Patterns are obvious and immediately recognizable
- MEDIUM: Patterns are clear but have minor variations/exceptions
- HARD: Patterns exist but require discovery (hidden regularity)
"""

from typing import List, Dict, Any, Tuple, Optional
import logging
import random
import copy

logger = logging.getLogger(__name__)


class PatternComplexityModifier:
    """
    Modifies pattern complexity for difficulty adaptation.
    
    Key principle: Difficulty is increased by making patterns harder to recognize,
    NOT by increasing item count or density.
    """
    
    def __init__(self):
        self._seed = None
    
    def set_seed(self, seed: int):
        """Set random seed for reproducible modifications."""
        self._seed = seed
        random.seed(seed)
    
    def apply_difficulty(
        self, 
        items: List[Dict[str, Any]], 
        difficulty_code: str,
        pattern_type: str = 'default'
    ) -> List[Dict[str, Any]]:
        """
        Apply difficulty-based pattern modifications.
        
        Args:
            items: List of item placements with 'type', 'pos', etc.
            difficulty_code: 'EASY', 'MEDIUM', or 'HARD'
            pattern_type: Optional pattern type hint from strategy
            
        Returns:
            Modified items list (same count, different arrangement/types)
        """
        if not items:
            return items
        
        difficulty = difficulty_code.upper() if difficulty_code else 'MEDIUM'
        
        # Deep copy to avoid mutating original
        modified_items = copy.deepcopy(items)
        
        if difficulty == 'EASY':
            modified_items = self._make_patterns_obvious(modified_items, pattern_type)
        elif difficulty == 'HARD':
            modified_items = self._obscure_patterns(modified_items, pattern_type)
        else:  # MEDIUM
            modified_items = self._add_minor_variations(modified_items, pattern_type)
        
        # Add pattern_complexity metadata
        for item in modified_items:
            item['pattern_complexity'] = difficulty
        
        original_count = len(items)
        final_count = len(modified_items)
        
        if original_count != final_count:
            logger.warning(f"Pattern modification changed item count: {original_count} → {final_count}")
            # Restore original count by trimming or duplicating
            if final_count > original_count:
                modified_items = modified_items[:original_count]
            elif final_count < original_count:
                # Duplicate last items to maintain count
                while len(modified_items) < original_count:
                    modified_items.append(copy.deepcopy(modified_items[-1]))
        
        logger.debug(f"Applied {difficulty} difficulty to {len(modified_items)} items")
        return modified_items
    
    def _make_patterns_obvious(
        self, 
        items: List[Dict[str, Any]], 
        pattern_type: str
    ) -> List[Dict[str, Any]]:
        """
        EASY: Make patterns immediately recognizable.
        
        Strategies:
        1. Ensure perfect regularity in spacing
        2. Group same item types together
        3. Use single pattern type only
        4. Remove any irregularities
        """
        if len(items) < 2:
            return items
        
        # Strategy 1: Ensure identical patterns per zone/branch
        zones = self._group_by_zone(items)
        
        if len(zones) > 1:
            # Make all zones have identical pattern
            reference_zone = list(zones.values())[0]
            reference_pattern = self._extract_pattern(reference_zone)
            
            for zone_name, zone_items in zones.items():
                self._apply_pattern(zone_items, reference_pattern)
        
        # Strategy 2: Regularize spacing
        items = self._regularize_spacing(items)
        
        # Strategy 3: Simplify item type sequence
        # E.g., ensure [C, C, C, S, S] instead of [C, S, C, S, C]
        items = self._group_item_types(items)
        
        # Mark as obvious
        for item in items:
            item['pattern_visibility'] = 'obvious'
        
        return items
    
    def _add_minor_variations(
        self, 
        items: List[Dict[str, Any]], 
        pattern_type: str
    ) -> List[Dict[str, Any]]:
        """
        MEDIUM: Add minor variations while keeping pattern clear.
        
        Strategies:
        1. Introduce 1-2 exceptions to the pattern
        2. Slight spacing variations (but still regular)
        3. One zone may differ from others
        """
        if len(items) < 3:
            return items
        
        # Strategy 1: Add one exception item
        items = self._add_pattern_exception(items)
        
        # Strategy 2: Vary one zone slightly
        zones = self._group_by_zone(items)
        if len(zones) > 2:
            # Pick one zone to have a slight variation
            zone_names = list(zones.keys())
            varied_zone = random.choice(zone_names[1:-1]) if len(zone_names) > 2 else zone_names[-1]
            
            zone_items = zones[varied_zone]
            if zone_items:
                self._introduce_variation(zone_items)
        
        # Mark as clear with variation
        for item in items:
            item['pattern_visibility'] = 'clear_with_variation'
        
        return items
    
    def _obscure_patterns(
        self, 
        items: List[Dict[str, Any]], 
        pattern_type: str
    ) -> List[Dict[str, Any]]:
        """
        HARD: Make patterns discoverable but not obvious.
        
        Strategies:
        1. Use non-uniform spacing with hidden regularity (e.g., Fibonacci)
        2. Mix item types in ways that hide the pattern
        3. Patterns span across zones (not within)
        4. Add noise that doesn't break the underlying pattern
        """
        if len(items) < 4:
            return items
        
        # Strategy 1: Apply progressive spacing
        items = self._apply_progressive_spacing(items)
        
        # Strategy 2: Interleave item types
        items = self._interleave_types(items)
        
        # Strategy 3: Add hidden regularity markers
        items = self._add_hidden_regularity(items)
        
        # Strategy 4: Mark one or more items as decoys
        items = self._mark_decoys(items)
        
        # Mark as hidden
        for item in items:
            item['pattern_visibility'] = 'hidden_regularity'
        
        return items
    
    # =========================================================================
    # HELPER METHODS
    # =========================================================================
    
    def _group_by_zone(self, items: List[Dict[str, Any]]) -> Dict[str, List[Dict[str, Any]]]:
        """Group items by their zone/branch/segment."""
        zones = {}
        for item in items:
            zone = item.get('zone') or item.get('pattern_id', '').split('_')[0] or 'default'
            if zone not in zones:
                zones[zone] = []
            zones[zone].append(item)
        return zones
    
    def _extract_pattern(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Extract the pattern template from a list of items."""
        pattern = []
        for i, item in enumerate(items):
            pattern.append({
                'index': i,
                'type': item.get('type', 'crystal'),
                'relative_pos': i  # Position relative to zone start
            })
        return pattern
    
    def _apply_pattern(self, items: List[Dict[str, Any]], pattern: List[Dict[str, Any]]):
        """Apply a pattern template to items."""
        for i, item in enumerate(items):
            if i < len(pattern):
                item['type'] = pattern[i]['type']
    
    def _regularize_spacing(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Ensure items have regular spacing (no modification to positions for EASY)."""
        # For EASY, we don't change positions, but we mark them as regular
        for i, item in enumerate(items):
            item['spacing_regularity'] = 'perfect'
        return items
    
    def _group_item_types(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Group same item types together for easier recognition."""
        # Count types
        types = {}
        for item in items:
            t = item.get('type', 'crystal')
            types[t] = types.get(t, 0) + 1
        
        # For EASY, we don't actually reorder (that would change positions),
        # but we mark the intended grouping
        for item in items:
            item['grouping'] = 'type_grouped'
        
        return items
    
    def _add_pattern_exception(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Add one exception to the pattern for MEDIUM difficulty."""
        if len(items) < 3:
            return items
        
        # Pick an item in the middle to be an exception
        exception_idx = len(items) // 2
        exception_item = items[exception_idx]
        
        # Mark it as exception
        exception_item['is_exception'] = True
        
        # Optionally swap its type
        types = ['crystal', 'switch']
        current_type = exception_item.get('type', 'crystal')
        if current_type in types:
            other_type = [t for t in types if t != current_type][0]
            exception_item['original_type'] = current_type
            exception_item['exception_type'] = other_type
            # Note: We don't actually change the type, just mark the intent
        
        return items
    
    def _introduce_variation(self, items: List[Dict[str, Any]]):
        """Introduce slight variation in a zone's pattern."""
        if not items:
            return
        
        # Mark as varied
        for item in items:
            item['has_variation'] = True
    
    def _apply_progressive_spacing(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Apply progressive spacing pattern (e.g., 1, 2, 3, 5, 8 - Fibonacci)."""
        for i, item in enumerate(items):
            # Mark the progressive nature
            item['spacing_pattern'] = 'progressive'
            item['position_in_sequence'] = i
            
            # Add Fibonacci-like hint
            fib = [1, 1, 2, 3, 5, 8, 13]
            if i < len(fib):
                item['spacing_hint'] = fib[i]
        
        return items
    
    def _interleave_types(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Interleave item types to hide simple patterns."""
        # For HARD, we mark items as interleaved but don't change actual types
        for i, item in enumerate(items):
            item['interleaving_index'] = i
            item['interleaving_type'] = 'type_a' if i % 2 == 0 else 'type_b'
        
        return items
    
    def _add_hidden_regularity(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Add markers showing hidden regularity pattern."""
        # Hidden pattern: every Nth item is special
        n = max(2, len(items) // 3)
        
        for i, item in enumerate(items):
            if i % n == 0:
                item['hidden_pattern_marker'] = True
                item['pattern_sequence'] = i // n
        
        return items
    
    def _mark_decoys(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Mark some items as decoys for HARD difficulty."""
        if len(items) < 5:
            return items
        
        # Mark ~20% of items as potential decoys
        num_decoys = max(1, len(items) // 5)
        decoy_indices = random.sample(range(len(items)), num_decoys)
        
        for i in decoy_indices:
            items[i]['is_potential_decoy'] = True
        
        return items
    
    # =========================================================================
    # VALIDATION
    # =========================================================================
    
    def validate_item_count_preserved(
        self, 
        original: List[Dict[str, Any]], 
        modified: List[Dict[str, Any]]
    ) -> bool:
        """Verify that modification preserved item count."""
        return len(original) == len(modified)
    
    def get_modification_summary(
        self, 
        items: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Get summary statistics of pattern modifications."""
        summary = {
            'total_items': len(items),
            'difficulty_levels': {},
            'visibility_types': {},
            'exceptions': 0,
            'decoys': 0,
        }
        
        for item in items:
            # Count by difficulty
            diff = item.get('pattern_complexity', 'unknown')
            summary['difficulty_levels'][diff] = summary['difficulty_levels'].get(diff, 0) + 1
            
            # Count by visibility
            vis = item.get('pattern_visibility', 'unknown')
            summary['visibility_types'][vis] = summary['visibility_types'].get(vis, 0) + 1
            
            # Count exceptions and decoys
            if item.get('is_exception'):
                summary['exceptions'] += 1
            if item.get('is_potential_decoy'):
                summary['decoys'] += 1
        
        return summary


# =============================================================================
# CONVENIENCE FUNCTIONS
# =============================================================================

def apply_difficulty_to_items(
    items: List[Dict[str, Any]], 
    difficulty_code: str
) -> List[Dict[str, Any]]:
    """
    Convenience function to apply difficulty modification.
    
    Args:
        items: Item list to modify
        difficulty_code: EASY, MEDIUM, or HARD
        
    Returns:
        Modified items with same count
    """
    modifier = PatternComplexityModifier()
    return modifier.apply_difficulty(items, difficulty_code)


def get_pattern_complexity_modifier() -> PatternComplexityModifier:
    """Get a PatternComplexityModifier instance."""
    return PatternComplexityModifier()

File: apps/map-builder-app/src/map-generator/handlers/PatternComplexityModifier.ts
/**
 * Pattern Complexity Modifier
 * 
 * Modifies item patterns based on difficulty level WITHOUT changing item count.
 * - EASY: Patterns are obvious and immediately recognizable
 * - MEDIUM: Patterns are clear but have minor variations/exceptions
 * - HARD: Patterns exist but require discovery (hidden regularity)
 * 
 * Ported from Python: pattern_complexity_modifier.py
 */

import { Coord } from '../types';

export interface ItemPlacement {
    type: string;
    pos?: Coord;
    position?: number[];
    pattern_id?: string;
    pattern_complexity?: string;
    pattern_visibility?: string;
    spacing_regularity?: string;
    grouping?: string;
    is_exception?: boolean;
    original_type?: string;
    exception_type?: string;
    has_variation?: boolean;
    spacing_pattern?: string;
    position_in_sequence?: number;
    spacing_hint?: number;
    interleaving_index?: number;
    interleaving_type?: string;
    hidden_pattern_marker?: boolean;
    pattern_sequence?: number;
    is_potential_decoy?: boolean;
    zone?: string;
    [key: string]: any;
}

/**
 * Pattern Complexity Modifier
 * Key principle: Difficulty is increased by making patterns harder to recognize,
 * NOT by increasing item count or density.
 */
export class PatternComplexityModifier {
    private seed: number | null = null;

    /**
     * Set random seed for reproducible modifications.
     */
    setSeed(seed: number): void {
        this.seed = seed;
    }

    /**
     * Apply difficulty-based pattern modifications.
     */
    applyDifficulty(
        items: ItemPlacement[],
        difficultyCode: string,
        patternType: string = 'default'
    ): ItemPlacement[] {
        if (!items || items.length === 0) {
            return items;
        }

        const difficulty = (difficultyCode || 'MEDIUM').toUpperCase();
        
        // Deep copy to avoid mutating original
        let modifiedItems = items.map(item => ({ ...item }));
        
        switch (difficulty) {
            case 'EASY':
                modifiedItems = this.makePatternsObvious(modifiedItems, patternType);
                break;
            case 'HARD':
                modifiedItems = this.obscurePatterns(modifiedItems, patternType);
                break;
            default: // MEDIUM
                modifiedItems = this.addMinorVariations(modifiedItems, patternType);
                break;
        }
        
        // Add pattern_complexity metadata
        modifiedItems.forEach(item => {
            item.pattern_complexity = difficulty;
        });
        
        // Ensure item count is preserved
        const originalCount = items.length;
        const finalCount = modifiedItems.length;
        
        if (originalCount !== finalCount) {
            console.warn(`[PatternComplexityModifier] Item count changed: ${originalCount} → ${finalCount}`);
            if (finalCount > originalCount) {
                modifiedItems = modifiedItems.slice(0, originalCount);
            } else {
                while (modifiedItems.length < originalCount) {
                    modifiedItems.push({ ...modifiedItems[modifiedItems.length - 1] });
                }
            }
        }
        
        console.log(`[PatternComplexityModifier] Applied ${difficulty} difficulty to ${modifiedItems.length} items`);
        return modifiedItems;
    }

    /**
     * EASY: Make patterns immediately recognizable.
     */
    private makePatternsObvious(items: ItemPlacement[], patternType: string): ItemPlacement[] {
        if (items.length < 2) return items;
        
        // Strategy 1: Ensure identical patterns per zone/branch
        const zones = this.groupByZone(items);
        
        if (Object.keys(zones).length > 1) {
            const referenceZone = Object.values(zones)[0];
            const referencePattern = this.extractPattern(referenceZone);
            
            for (const zoneItems of Object.values(zones)) {
                this.applyPattern(zoneItems, referencePattern);
            }
        }
        
        // Strategy 2: Regularize spacing
        items = this.regularizeSpacing(items);
        
        // Strategy 3: Group item types
        items = this.groupItemTypes(items);
        
        // Mark as obvious
        items.forEach(item => {
            item.pattern_visibility = 'obvious';
        });
        
        return items;
    }

    /**
     * MEDIUM: Add minor variations while keeping pattern clear.
     */
    private addMinorVariations(items: ItemPlacement[], patternType: string): ItemPlacement[] {
        if (items.length < 3) return items;
        
        // Strategy 1: Add one exception item
        items = this.addPatternException(items);
        
        // Strategy 2: Vary one zone slightly
        const zones = this.groupByZone(items);
        const zoneNames = Object.keys(zones);
        
        if (zoneNames.length > 2) {
            const variedZone = zoneNames.length > 2 
                ? zoneNames[Math.floor(zoneNames.length / 2)]
                : zoneNames[zoneNames.length - 1];
            
            const zoneItems = zones[variedZone];
            if (zoneItems && zoneItems.length > 0) {
                this.introduceVariation(zoneItems);
            }
        }
        
        // Mark as clear with variation
        items.forEach(item => {
            item.pattern_visibility = 'clear_with_variation';
        });
        
        return items;
    }

    /**
     * HARD: Make patterns discoverable but not obvious.
     */
    private obscurePatterns(items: ItemPlacement[], patternType: string): ItemPlacement[] {
        if (items.length < 4) return items;
        
        // Strategy 1: Apply progressive spacing
        items = this.applyProgressiveSpacing(items);
        
        // Strategy 2: Interleave item types
        items = this.interleaveTypes(items);
        
        // Strategy 3: Add hidden regularity markers
        items = this.addHiddenRegularity(items);
        
        // Strategy 4: Mark decoys
        items = this.markDecoys(items);
        
        // Mark as hidden
        items.forEach(item => {
            item.pattern_visibility = 'hidden_regularity';
        });
        
        return items;
    }

    // =========================================================================
    // HELPER METHODS
    // =========================================================================

    private groupByZone(items: ItemPlacement[]): Record<string, ItemPlacement[]> {
        const zones: Record<string, ItemPlacement[]> = {};
        
        for (const item of items) {
            const zone = item.zone || 
                        (item.pattern_id ? item.pattern_id.split('_')[0] : 'default');
            
            if (!zones[zone]) {
                zones[zone] = [];
            }
            zones[zone].push(item);
        }
        
        return zones;
    }

    private extractPattern(items: ItemPlacement[]): { index: number; type: string; relativePos: number }[] {
        return items.map((item, i) => ({
            index: i,
            type: item.type || 'crystal',
            relativePos: i
        }));
    }

    private applyPattern(
        items: ItemPlacement[], 
        pattern: { index: number; type: string; relativePos: number }[]
    ): void {
        for (let i = 0; i < items.length && i < pattern.length; i++) {
            items[i].type = pattern[i].type;
        }
    }

    private regularizeSpacing(items: ItemPlacement[]): ItemPlacement[] {
        items.forEach(item => {
            item.spacing_regularity = 'perfect';
        });
        return items;
    }

    private groupItemTypes(items: ItemPlacement[]): ItemPlacement[] {
        items.forEach(item => {
            item.grouping = 'type_grouped';
        });
        return items;
    }

    private addPatternException(items: ItemPlacement[]): ItemPlacement[] {
        if (items.length < 3) return items;
        
        const exceptionIdx = Math.floor(items.length / 2);
        const exceptionItem = items[exceptionIdx];
        
        exceptionItem.is_exception = true;
        
        const types = ['crystal', 'switch'];
        const currentType = exceptionItem.type || 'crystal';
        
        if (types.includes(currentType)) {
            const otherType = types.find(t => t !== currentType) || 'crystal';
            exceptionItem.original_type = currentType;
            exceptionItem.exception_type = otherType;
        }
        
        return items;
    }

    private introduceVariation(items: ItemPlacement[]): void {
        items.forEach(item => {
            item.has_variation = true;
        });
    }

    private applyProgressiveSpacing(items: ItemPlacement[]): ItemPlacement[] {
        const fibonacci = [1, 1, 2, 3, 5, 8, 13];
        
        items.forEach((item, i) => {
            item.spacing_pattern = 'progressive';
            item.position_in_sequence = i;
            
            if (i < fibonacci.length) {
                item.spacing_hint = fibonacci[i];
            }
        });
        
        return items;
    }

    private interleaveTypes(items: ItemPlacement[]): ItemPlacement[] {
        items.forEach((item, i) => {
            item.interleaving_index = i;
            item.interleaving_type = i % 2 === 0 ? 'type_a' : 'type_b';
        });
        
        return items;
    }

    private addHiddenRegularity(items: ItemPlacement[]): ItemPlacement[] {
        const n = Math.max(2, Math.floor(items.length / 3));
        
        items.forEach((item, i) => {
            if (i % n === 0) {
                item.hidden_pattern_marker = true;
                item.pattern_sequence = Math.floor(i / n);
            }
        });
        
        return items;
    }

    private markDecoys(items: ItemPlacement[]): ItemPlacement[] {
        if (items.length < 5) return items;
        
        const numDecoys = Math.max(1, Math.floor(items.length / 5));
        const indices = this.randomSample(items.length, numDecoys);
        
        indices.forEach(i => {
            items[i].is_potential_decoy = true;
        });
        
        return items;
    }

    private randomSample(max: number, count: number): number[] {
        const indices: number[] = [];
        const available = Array.from({ length: max }, (_, i) => i);
        
        for (let i = 0; i < count && available.length > 0; i++) {
            const randomIdx = Math.floor(Math.random() * available.length);
            indices.push(available[randomIdx]);
            available.splice(randomIdx, 1);
        }
        
        return indices;
    }

    // =========================================================================
    // VALIDATION
    // =========================================================================

    /**
     * Verify that modification preserved item count.
     */
    validateItemCountPreserved(original: ItemPlacement[], modified: ItemPlacement[]): boolean {
        return original.length === modified.length;
    }

    /**
     * Get summary statistics of pattern modifications.
     */
    getModificationSummary(items: ItemPlacement[]): Record<string, any> {
        const summary: Record<string, any> = {
            total_items: items.length,
            difficulty_levels: {},
            visibility_types: {},
            exceptions: 0,
            decoys: 0
        };
        
        for (const item of items) {
            // Count by difficulty
            const diff = item.pattern_complexity || 'unknown';
            summary.difficulty_levels[diff] = (summary.difficulty_levels[diff] || 0) + 1;
            
            // Count by visibility
            const vis = item.pattern_visibility || 'unknown';
            summary.visibility_types[vis] = (summary.visibility_types[vis] || 0) + 1;
            
            // Count exceptions and decoys
            if (item.is_exception) summary.exceptions++;
            if (item.is_potential_decoy) summary.decoys++;
        }
        
        return summary;
    }
}

// Singleton instance
let modifierInstance: PatternComplexityModifier | null = null;

export function getPatternComplexityModifier(): PatternComplexityModifier {
    if (!modifierInstance) {
        modifierInstance = new PatternComplexityModifier();
    }
    return modifierInstance;
}

/**
 * Convenience function to apply difficulty modification.
 */
export function applyDifficultyToItems(
    items: ItemPlacement[],
    difficultyCode: string
): ItemPlacement[] {
    const modifier = new PatternComplexityModifier();
    return modifier.applyDifficulty(items, difficultyCode);
}
